torch/csrc/autograd/functions/accumulate_grad.h

```diff
@@ -110,9 +171,33 @@ struct TORCH_API AccumulateGrad : public Node {
         // place. `variable_grad` is thus still referring to the same tensor
         // after the operation.
         variable_grad += new_grad;
+        // ^ We could enforce the contract more aggressively here by writing:
+        // if (variable_grad.is_sparse() || new_grad.is_sparse()) {
+        //   variable_grad += new_grad;
+        // } else if (obeys_layout_contract(variable_grad, variable)) {
+        //   variable_grad += new_grad;
+        // } else {
+        //   result = at::empty_strided(variable.sizes(), variable.strides(),
+        //                              variable.options().memory_format(c10::nullopt));
+        //   update_grad(at::native::add_out(result, variable_grad, new_grad, 1.0);
+        // }
+        // However, that accumulation is sometimes in place and sometimes not,
+        // which may break user code.
       }
     } else {
-      update_grad(variable_grad + new_grad);
+      // Assumes operator+ result typically matches strides of first arg.
+      if (obeys_layout_contract(new_grad, variable)) {
+        update_grad(new_grad + variable_grad);
+      } else {
+        update_grad(variable_grad + new_grad);
```

Comment:
Why not always do this (compared to new_grad + variable_grad above)? Here we assume that the `variable_grad` already follows the contract (or is not because the user explicitly created it not to).
This would remove a check at runtime and slightly simplify this code.

Reply:
agreed after talking on slack